photon-client/src/main/scala/com/linkedin/photon/ml/io/scopt/ScoptParameter.scala

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright 2017 LinkedIn Corp. All rights reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linkedin.photon.ml.io.scopt
+
+import org.apache.spark.ml.param.{Param, ParamMap}
+import scopt.{OptionDef, Read}
+
+/**
+ * A Scopt wrapper around a [[Param]] for parsing it from the command line, and writing it out in the same format.
+ *
+ * @tparam In The type of the command line argument when parsed by Scopt
+ * @tparam Out The type of the underlying [[Param]]
+ * @param param The base [[Param]] being wrapped
+ * @param parse The function to parse the parameter value from a basic Scopt type
+ * @param update The function to update the value of the parameter in the [[ParamMap]] if it's specified multiple times
+ * @param print The function to print multiple parameter values in Scopt-parseable format
+ * @param usageText Usage text for the parameter
+ * @param additionalDocs Additional documentation, in addition to the documentation of the base parameter
+ * @param isRequired Whether the parameter must be included within the command line args
+ * @param isUnbounded Whether the parameter can occur an unlimited number of times
+ * @param read An implicit [[scopt.Read]] to parse the command line argument
+ */
+class ScoptParameter[In, Out] private (
+    param: Param[Out],
+    parse: (In) => Out,
+    update: (Out, Out) => Out,
+    print: (Out) => Seq[String],
+    usageText: String,
+    additionalDocs: Seq[String],
+    isRequired: Boolean,
+    isUnbounded: Boolean)
+    (implicit val read: Read[In]) {
```

Comment:
@joshvfleming 
@fastier-li 
@Adama94

A wrapper around a `spark.ml.param.Param` for storing all Scopt command line parsing/printing info, as well as usage and additional documentation.